File: experiments/java-sdk/src/main/java/com/datadog/llmobs/experiments/ExperimentResult.java
package com.datadog.llmobs.experiments;

import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * Result of an {@link Experiment#run()}.
 *
 * <p>Each {@link Row} carries the per-record input, output, error, and the evaluator results.
 * {@link #url()} points at the experiment in the Datadog dashboard.
 */
public final class ExperimentResult {

    private final String experimentId;
    private final List<Row> rows;
    private final String url;

    ExperimentResult(String experimentId, List<Row> rows, String url) {
        this.experimentId = experimentId;
        this.rows = Collections.unmodifiableList(rows);
        this.url = url;
    }

    public String experimentId() {
        return experimentId;
    }

    public List<Row> rows() {
        return rows;
    }

    public String url() {
        return url;
    }

    /** Per-record result. */
    public static final class Row {
        private final int index;
        private final String spanId;
        private final String traceId;
        private final long startNs;
        private final long durationNs;
        private final Object input;
        private final Object output;
        private final Object expectedOutput;
        private final String errorType;
        private final String errorMessage;
        private final Map<String, Object> evaluations;
        private final Map<String, String> evaluationErrors;

        Row(int index, String spanId, String traceId, long startNs, long durationNs,
            Object input, Object output, Object expectedOutput,
            String errorType, String errorMessage,
            Map<String, Object> evaluations, Map<String, String> evaluationErrors) {
            this.index = index;
            this.spanId = spanId;
            this.traceId = traceId;
            this.startNs = startNs;
            this.durationNs = durationNs;
            this.input = input;
            this.output = output;
            this.expectedOutput = expectedOutput;
            this.errorType = errorType;
            this.errorMessage = errorMessage;
            this.evaluations = Collections.unmodifiableMap(evaluations);
            this.evaluationErrors = Collections.unmodifiableMap(evaluationErrors);
        }

        public int index() { return index; }
        public String spanId() { return spanId; }
        public String traceId() { return traceId; }
        public long startNs() { return startNs; }
        public long durationNs() { return durationNs; }
        public Object input() { return input; }
        public Object output() { return output; }
        public Object expectedOutput() { return expectedOutput; }
        public boolean isError() { return errorType != null; }
        public String errorType() { return errorType; }
        public String errorMessage() { return errorMessage; }
        public Map<String, Object> evaluations() { return evaluations; }
        public Map<String, String> evaluationErrors() { return evaluationErrors; }
    }
}

File: experiments/node-sdk/src/ExperimentResult.ts
/**
 * One row of an experiment run — mirrors the Java `ExperimentResult.Row`.
 *
 * Implemented as a class (rather than a plain object) so `isError` is a derived
 * accessor, matching the Java getter `isError()`.
 */
export class Row {
  readonly index: number;
  readonly spanId: string;
  readonly traceId: string;
  readonly startNs: number;
  readonly durationNs: number;
  readonly input: unknown;
  readonly output: unknown;
  readonly expectedOutput: unknown;
  readonly errorType: string | null;
  readonly errorMessage: string | null;
  readonly evaluations: Record<string, unknown>;
  readonly evaluationErrors: Record<string, string>;

  constructor(args: {
    index: number;
    spanId: string;
    traceId: string;
    startNs: number;
    durationNs: number;
    input: unknown;
    output: unknown;
    expectedOutput: unknown;
    errorType: string | null;
    errorMessage: string | null;
    evaluations: Record<string, unknown>;
    evaluationErrors: Record<string, string>;
  }) {
    this.index = args.index;
    this.spanId = args.spanId;
    this.traceId = args.traceId;
    this.startNs = args.startNs;
    this.durationNs = args.durationNs;
    this.input = args.input;
    this.output = args.output;
    this.expectedOutput = args.expectedOutput;
    this.errorType = args.errorType;
    this.errorMessage = args.errorMessage;
    this.evaluations = args.evaluations;
    this.evaluationErrors = args.evaluationErrors;
  }

  get isError(): boolean {
    return this.errorType !== null;
  }
}

/** Returned by `Experiment.run()` — mirrors the Java `ExperimentResult`. */
export class ExperimentResult {
  readonly experimentId: string;
  readonly rows: Row[];
  readonly url: string;

  constructor(experimentId: string, rows: Row[], url: string) {
    this.experimentId = experimentId;
    this.rows = rows;
    this.url = url;
  }
}
